### ISSUE TITLE
Pass slack.Msg to commands

### ISSUE THREAD

**Issue by pasali** [lnxjedi/gopherbot#9]
There is a lot of information in slack.Msg like file, attachment, comment etc. so passing it in commands would be very nice. Just passing slack.Msg.Text won't be enough in every case.
For ex: 
![](https://slack.atlassian.io/img/faq-14.png)

**Comment by parsley42** [comment]
My intention is that Gopherbot be able to support multiple protocols, so I'm trying not to make any design decisions that tie it completely to slack. I'm open, however, to protocol-specific extensions, if implemented cleanly. Also keep in mind it's designed to support multiple languages for plugins - while you primarily use Go plugins, I write plugins in Python, Ruby, PowerShell, Bash and Go - so the plugin calling convention is fairly fixed to something that works well when calling a command-line script.

Basically, the current design doesn't really support passing in a complex structure to the plugin. To support this kind of functionality in a script-friendly and protocol-agnostic way, it would mean allowing custom, protocol-specific matching functions. For instance, eventually there will be events for presence change, attachment received, etc.

I'll give this some more thought.

**Comment by pasali** [comment]
I understand your point. I also think that you will need this functionality at some point. Because eventually people are going to want these informations :).

**Comment by parsley42** [comment]
Ok, I think I've addressed this for you in 69101ac, on the plugdev branch. Once I've filled out the plugin debugging code I'll merge this to master; once I've done more testing I'll go ahead and release v1.1.0, probably next week.

Your Go plugin should check bot.Protocol == bot.Slack, then use a type assertion to get the *slack.MessageEvent from bot.RawMsg. **NOTE:** I have *not* tested this AT ALL, though I intend to soon. Should be straightforward coding, but it's always possible I screwed something up. (although I've written a test suite now so I'm fairly confident I haven't screwed anything up badly)

**Comment by pasali** [comment]
thanks. i will give a try when i find a chance

**Comment by pasali** [comment]
@parsley42 MessageMatchers are not working now. When message received, it checks whether is command to bot or just text to channel
```go	
if preRegex != nil {
		matches := preRegex.FindAllStringSubmatch(messageFull, -1)
		if matches != nil && len(matches[0]) == 2 {
			isCommand = true
			message = matches[0][1]
		}
}
```

but preRegex is something like that `^(?i:|@?joe[:, ]\s*)(.*)$` which matches even robot was not mentioned. So below code block does not get executed. Because `bot.isCommnd` is `true`

```go
	if !commandMatched && !waitingForReply && !bot.isCommand {
		// check for ambient message matches
		commandMatched = bot.checkPluginMatchersAndRun(false)
	}
```

**Comment by parsley42** [comment]
Ouch.

Ok, I'll fix this and add tests for it, and get back with you.

**Comment by parsley42** [comment]
Not sure what version you were using, but that was an old version of the regex, I think. In any case, older versions of Gopherbot wouldn't use MessageMatchers for DMs, but I've corrected that, as well. The new Robot object carries more info and allows this change (see the most recent commit).

So I added a test plugin that uses MessageMatchers, and added tests to the test suite - it all looks good now. Care to give it a shot with the latest snapshot / commit?

**Comment by parsley42** [comment]
Ok, I took a little more time to look at this and found the issue - looking at your regex, I realized you hadn't configured an 'alias' for your bot. I fixed the regex generating algorithm and added tests for correct behavior when the robot has either no name or no alias.

Additionally, after thinking about it I realized the robot should check MessageMatchers if the user sends the robot a command, but it doesn't match. So, 'hello robot' and 'robot, hello robot' now both trigger the 'hello robot' MessageMatcher (unless there's a 'hello robot' command). (... and of course, my new mantra - 'I've added tests for this behavior').

All of these fixes are on master for v1.1.0-snapshot. I'm planning on fleshing out test coverage a bit more then releasing v1.1.0. Thanks for testing.

**Comment by pasali** [comment]
the last version you have released is OK for me now. I am done with development and its works like expected. We can close this issue by the time you release production version. Thanks for the feature.

**Comment by parsley42** [comment]
Good to hear, thanks.

**Comment by parsley42** [comment]
Version 1.1.0 has been released with this fix.
